### DISCUSSION TITLE
Does the client update URLs for a cluster automatically?

### DISCUSSION THREAD

**Discussion by damccull** [ravendb/ravendb#14975] [Q&A] [answered]
I see there's a `/cluster/topology` endpoint in the REST api. Does the client make use of this to update the server URLs automatically after the DocumentStore is created?

If so, how often and what operations might trigger it? I do see that it will fetch the topology on the first database access, as noted in the DocumentStore documentation, but does it continue to keep it up to date if the topology changes externally to the client?

 Thanks.

**Comment by ayende** [comment, marked as answer]
Yes, the client will get the new URLs from the server ( and cache them locally ).
It happens on the next request after the topology change.

**Reply by damccull** [reply]
Great! Thank you. What is the signal that the topology has changed? Does the server send a notification of a change via the response metadata or headers or similar? Or does the client just download the topology automatically as part of every request?

**Reply by arekpalinski** [reply]
The client is notified by `Refresh-Topology: true` response header

**Reply by damccull** [reply]
> The client is notified by `Refresh-Topology: true` response header

Awesome! Thank you. That's a big help!

**Reply by damccull** [reply]
I'm trying to trigger this header manually in postman by changing the topology of my local dev cluster. However, it's never showing up. I'm sending `http://localhost:8080/databases/sample/docs?pageSize=4`, viewing the headers, removing or adding a node, re-sending the same request, the comparing the new response headers. I never see `Refresh-Topology` as a header. How do I trigger this?

Thinking about it, I realize the db may not want to keep track of every source IP the database has already notified, so perhaps there's some way to register myself as a client to be notified of topology refreshes that I haven't found yet?

**Reply by ayende** [reply]
Are you adding a node from the _database_, or from the cluster?
And in order to see that, you should tell the server what version of the topology you have.

We are only sending that for requests that specify `Topology-Etag`

**Reply by damccull** [reply]
I am adding/deleting the node in the cluster management. I'll try again from the database. I will also try sending with a topology etag. Thanks.

**Reply by damccull** [reply]
I was able to get the header after sending the Topology-Etag header with the current etag I'd stored, and on the query `http://localhost:8080/databases/sample/docs?pageSize=5`. So I'm good there now.

Further question, however, is his long will the server continue to send that response header? Repeated requests, even after updating the Etag with new one seems to result in getting the `Refresh-Topology` response header every request for 5 or so minutes.

Is this a caching issue or is the server continuing to send it for a while?

**Reply by ayende** [reply]
Until you send the request with the right topology etag.

**Reply by damccull** [reply]
Ah, ok. So if the Etag updates to 65, and I send 63, I'll start seeing the header. Then I need to send a request with the Etag set to 65 before I stop seeing it, as I now understand. Is this condition solved by any request that includes the correct Etag or just the /cluster/topology request containing the right Etag?

Thanks for all your help on this, by the way.

**Reply by ayende** [reply]
This is done on a _per request_ basis, there is no state that the server is tracking.

**Reply by damccull** [reply]
Ok, great to know! I think I successfully understand. I appreciate your help and I hope this thread will be useful to others in the future too.

**Reply by damccull** [reply]
I see strange behavior which leads me to think I'm not fully understanding this yet. Even if I send `Topology-Etag: 0` as a header, I do not see the `Refresh-Topology: true` response header unless I first make a topology change. Then, no matter how many times I send the correct topology etag back, I will keep getting the Refresh-Topology response header for about 5 minutes or so. I don't have any caching going on in my program that I'm aware of. I certainly haven't added any. Am I still doing this wrong?

**Reply by arekpalinski** [reply]
Can you share any piece of code that you have?

**Reply by damccull** [reply]
Sure: https://github.com/damccull/ravendb_client/blob/main/ravendb_client/src/document_store/document_store_actor.rs#L235
This is where I am adding the header to my outgoing request. This is in rust and very early, non-optimal code, so I apologize for the probable rough read.

I will continue to troubleshoot this on my end as well. I'm seeing it in postman with manually produced requests as well.

**Reply by ppekrol** [reply]
After you receive the 'True' in Refresh-Topology header, you need to fetch latest topology (and update the etag) yourself. The code in C# is here: https://github.com/ravendb/ravendb/blob/cd0ce04c44454272d85ccf75782a6be6b3bff7fc/src/Raven.Client/Http/RequestExecutor.cs#L921-L945
and is executed after each request here: https://github.com/ravendb/ravendb/blob/cd0ce04c44454272d85ccf75782a6be6b3bff7fc/src/Raven.Client/Http/RequestExecutor.cs#L870

Do you have sth similar there? I couldn't find it.

**Reply by ayende** [reply]
@damccull If you are building a Rust client for RavenDB, I would recommend that you'll follow the structure of the other clients.
(Request executor, commands, operations, etc).

That tends to be the best way to achieve parity in our experiene.

**Reply by damccull** [reply]
> After you receive the 'True' in Refresh-Topology header, you need to fetch latest topology (and update the etag) yourself. The code in C# is here:
> 
> https://github.com/ravendb/ravendb/blob/cd0ce04c44454272d85ccf75782a6be6b3bff7fc/src/Raven.Client/Http/RequestExecutor.cs#L921-L945
> 
> 
> and is executed after each request here:
> https://github.com/ravendb/ravendb/blob/cd0ce04c44454272d85ccf75782a6be6b3bff7fc/src/Raven.Client/Http/RequestExecutor.cs#L870
> 
> Do you have sth similar there? I couldn't find it.

This function is kicked off to update the topology in an async task if the Refresh-Topology header exists and is "true": https://github.com/damccull/ravendb_client/blob/main/ravendb_client/src/document_store/document_store_actor.rs#L129
It stores the task's handle in a field on the DocumentStoreActor while the task runs in the background. It'll only kick off if another one is not already running.
Once the topology download task is completed, it is currently loaded into the correct DocumentStoreActor field in the next call to get_server_address() as the first thing done, if a topology update task has been launched and then completed.
https://github.com/damccull/ravendb_client/blob/main/ravendb_client/src/document_store/document_store_actor.rs#L247

I intend to combine these two separate operations into more of a callback style so the updated topology is stored for use as soon as it's download is complete, so that will be changing for the better.

So, I am storing the received topology back, and subsequent requests after it's stored are using the updated topology addresses. I know this because my first test request in the 'demo' example in the same repo will use "localhost:8000" while requests made after the update (the second request, after a 2-second thread sleep) will use "raven1:8000" or one of its siblings.

Here is an example of the problem I am seeing in Postman, where I am able to reliably replicate it:
I run this query to get the topology etag:
![image](https://user-images.githubusercontent.com/401104/193038164-7756ca2f-bbc0-4189-ac11-518c1c005072.png)

Then I run this query with the etag set to 60 to match what came back in the topology, but it still has the Refresh-Topology=true header.
![image](https://user-images.githubusercontent.com/401104/193038672-accb5cd8-773b-4b22-86a0-45ec5912c250.png)

This issue isn't affecting my ability to get data or know the correct topology, but it is causing me to send needless requests to the server to update the topology every time, even if I send it the current etag. That's why I'm wanting to know if I've got a bad understanding of the behavior.

**Reply by ayende** [reply]
What time zone are you on? Might be easier to do that in a meeting.

**Reply by damccull** [reply]
> #

Mr. @ayende, 

I am doing my best to replicate the existing clients, both in user experience as well as in how they function. The nature of the rust language means I have to make some different decisions in some areas, though. Where the C# and Java clients send direct references to the DocumentStore class all over to various other objects so they can reference data on it, I'm using an actor with a cloneable handle and messaging to accomplish the same. This is to provide the same singleton-like behavior the C# client has, but in a more rust-idiomatic way and while hoping to avoid atomic reference counters and mutexes.

However, I am already using a command struct to send the actual commands.
In https://github.com/damccull/ravendb_client/blob/main/ravendb_client/src/raven_command.rs you can see this command struct.

It's currently lightweight but will gain more variants in the enum near the bottom over time. The idea is to have methods the user calls, like in C# client, then the library will build a RavenCommand and pass it to an executor to actually get sent to the server and do the work.

I'm trying to understand the existing clients as I go and making good headway, but I've only just started my project as a hobby and it's definitely going to need some reworking as I go. I appreciate your advice. My goal is to make a usable library for myself that operates, at least from a user perspective, as closely to the C# library as possible. I will gladly take any advice the RavenDB devs have for me, though.

**Reply by damccull** [reply]
> What time zone are you on? Might be easier to do that in a meeting.

Oh, I'm in EST. I'm not in a rush to finish this. It'd be nice if I can get it to a basic working configuration for sending, receiving, and mutating documents by next year. I'm quite new to projects of this scale, and I barely understand how the overall system will work yet ;). This is a great learning opportunity for me. So I would be willing to talk with you, of course, but I definitely don't want to waste your time.

**Reply by ayende** [reply]
Would be happy to do a call with you next week, if you would like. 
It's a pleasure to help such an endeavor.

**Reply by damccull** [reply]
Gentlemen, I believe I've figured out the wacky behavior I've been experiencing. Digging back into the docs again, I realized that both the cluster _and_ the database have separate topologies. It seems that I misunderstood how the topology etag is used.

I now see that sending a Topology-Etag that matches the cluster will result in not getting a `Refresh-Topology` header for the `/cluster/topology` endpoint, but _will still_ result in the `Refresh-Topology:true` header for any calls to endpoints for the database itself, if the database topology etag is higher than the cluster's.

Now, knowing about this, I am no longer seeing the weird behavior described above.

**Reply by ayende** [reply]
Oh, sorry about that. Yeah, topology exists in different layers, we didn't consider the terminology here.
